README
pam_fprint
==========

Homepage: http://www.reactivated.net/fprint/wiki/Pam_fprint

pam_fprint is a PAM module which uses libfprint's fingerprint verification
functionality to implement a fingerprint-based login system.

Before you can use pam_fprint to login, you must have enrolled at least one
finger on the device that you wish to use (or a compatible one).
A program called pam_fprint_enroll is provided that does this for you.
Run with the --help argument to learn how to enroll different fingers.

pam_fprint is alpha software and has deficiencies.

See the the homepage for PAM setup instructions and other notes.

Support and development happens on the mailing list:
http://www.reactivated.net/fprint/wiki/Mailing_list

Enjoy!
 - Daniel Drake <dsd@gentoo.org>


NEWS
This file lists notable changes in each release. For the full history of all
changes, see ChangeLog.

2007-11-19: v0.2 release
 * Don't try to serve sessions from remote hosts (Daniel Drake)
 * Add enrollment application (Vasily Khoruzhick)

2007-11-15: v0.1 release
 * Initial public release



src/pamtest.c
/***
  This file is part of pam_dotfile.

  pam_dotfile is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.
  
  pam_dotfile is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with pam_dotfile; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
  USA
***/

#include <stdio.h>

#include <security/pam_appl.h>
#include <security/pam_misc.h>

int main(int argc, char*argv[]) {
    static struct pam_conv pc = { misc_conv, NULL };
    pam_handle_t *ph = NULL;
    int r, ret;
    char *username, *procname, *service;

    if ((procname = strchr(argv[0], '/')))
        procname++;
    else
        procname = argv[0];

    if (argc <= 1 || argc > 3) {
        fprintf(stderr, "Usage: %s [<service>] [<username>]\n", procname);
        exit(1);
    }

    service = (argc >= 2) ? argv[1] : procname;
    username = (argc == 3) ? argv[2] : NULL;

    if (username)
        printf("Trying to authenticate <%s> for service <%s>.\n", username, service);
    else
        printf("Trying to authenticate for service <%s>.\n", service);
    
    if ((r = pam_start(service, username, &pc, &ph)) != PAM_SUCCESS) {
        fprintf(stderr, "Failure starting pam: %s\n", pam_strerror(ph, r));
        return 1;
    }

    if ((r = pam_authenticate(ph, 0)) != PAM_SUCCESS) {
        fprintf(stderr, "Failed to authenticate: %s\n", pam_strerror(ph, r));
        ret = 1;
    } else {
        printf("Authentication successful.\n");
        ret = 0;
    }

    if ((r = pam_end(ph, r)) != PAM_SUCCESS)
        fprintf(stderr, "Failure shutting down pam: %s\n", pam_strerror(ph, r));

    return ret;
}


src/pam_fprint.c
/*
 * pam_fprint: PAM module for fingerprint authentication through libfprint
 * Copyright (C) 2007 Daniel Drake <dsd@gentoo.org>
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <stdio.h>
#include <unistd.h>
#include <sys/types.h>
#include <pwd.h>
#include <string.h>

#include <fprint.h>

#define PAM_SM_AUTH
#include <security/pam_modules.h>

static int send_info_msg(pam_handle_t *pamh, char *msg)
{
	const struct pam_message mymsg = {
		.msg_style = PAM_TEXT_INFO,
		.msg = msg,
	};
	const struct pam_message *msgp = &mymsg;
    const struct pam_conv *pc;
	struct pam_response *resp;
	int r;

    r = pam_get_item(pamh, PAM_CONV, (const void **) &pc);
	if (r != PAM_SUCCESS)
		return;

	if (!pc || !pc->conv)
		return;

	return pc->conv(1, &msgp, &resp, pc->appdata_ptr);
}

static int send_err_msg(pam_handle_t *pamh, char *msg)
{
	const struct pam_message mymsg = {
		.msg_style = PAM_ERROR_MSG,
		.msg = msg,
	};
	const struct pam_message *msgp = &mymsg;
    const struct pam_conv *pc;
	struct pam_response *resp;
	int r;

    r = pam_get_item(pamh, PAM_CONV, (const void **) &pc);
	if (r != PAM_SUCCESS)
		return;

	if (!pc || !pc->conv)
		return;

	return pc->conv(1, &msgp, &resp, pc->appdata_ptr);
}

static const char *fingerstr(enum fp_finger finger)
{
	const char *names[] = {
		[LEFT_THUMB] = "left thumb",
		[LEFT_INDEX] = "left index",
		[LEFT_MIDDLE] = "left middle",
		[LEFT_RING] = "left ring",
		[LEFT_LITTLE] = "left little",
		[RIGHT_THUMB] = "right thumb",
		[RIGHT_INDEX] = "right index",
		[RIGHT_MIDDLE] = "right middle",
		[RIGHT_RING] = "right ring",
		[RIGHT_LITTLE] = "right little",
	};
	if (finger < LEFT_THUMB || finger > RIGHT_LITTLE)
		return "UNKNOWN";
	return names[finger];
}

static int find_dev_and_print(struct fp_dscv_dev **ddevs,
	struct fp_dscv_print **prints, struct fp_dscv_dev **_ddev,
	struct fp_dscv_print **_print)
{
	int i = 0;
	struct fp_dscv_print *print;
	struct fp_dscv_dev *ddev;

	while (print = prints[i++]) {
		ddev = fp_dscv_dev_for_dscv_print(ddevs, print);
		if (ddev) {
			*_ddev = ddev;
			*_print = print;
			return 0;
		}
	}
	return 1;
}

static int do_verify(pam_handle_t *pamh, struct fp_dev *dev,
	struct fp_print_data *data, enum fp_finger finger)
{
	int max_tries = 5;
	const char *driver_name = fp_driver_get_full_name(fp_dev_get_driver(dev));
	const char *fstr = fingerstr(finger);

	do {
		int r;
		char msg[128];

		snprintf(msg, sizeof(msg), "Scan %s finger on %s", fstr, driver_name);
		msg[sizeof(msg) - 1] = 0;
		send_info_msg(pamh, msg);

		r = fp_verify_finger(dev, data);
		if (r < 0) {
			snprintf(msg, sizeof(msg), "Fingerprint verification error %d", r);
			msg[sizeof(msg) - 1] = 0;
			send_err_msg(pamh, msg);
			return PAM_AUTHINFO_UNAVAIL;
		}
		switch (r) {
		case FP_VERIFY_NO_MATCH:
			return PAM_AUTH_ERR;
		case FP_VERIFY_MATCH:
			return PAM_SUCCESS;
		case FP_VERIFY_RETRY:
			send_info_msg(pamh, "Scan didn't quite work. Please try again.");
			break;
		case FP_VERIFY_RETRY_TOO_SHORT:
			send_info_msg(pamh, "Swipe was too short, please try again.");
			break;
		case FP_VERIFY_RETRY_CENTER_FINGER:
			send_info_msg(pamh, "Please center your finger on the sensor and "
				"try again.");
			break;
		case FP_VERIFY_RETRY_REMOVE_FINGER:
			send_info_msg(pamh, "Please remove finger from the sensor and try "
				"again.");
			break;
		}
	} while (max_tries--);

	send_err_msg(pamh, "Too many failed scans, giving up.");
	return PAM_AUTHINFO_UNAVAIL;
}

static int do_auth(pam_handle_t *pamh)
{
	int r;
	struct fp_dscv_dev **ddevs;
	struct fp_dscv_print **prints;
	struct fp_dscv_dev *ddev;
	struct fp_dscv_print *print;
	struct fp_dev *dev;
	struct fp_print_data *data;
	enum fp_finger finger;

	r = fp_init();
	if (r < 0)
		return PAM_AUTHINFO_UNAVAIL;

	ddevs = fp_discover_devs();
	if (!ddevs)
		return PAM_AUTHINFO_UNAVAIL;

	prints = fp_discover_prints();
	if (!prints) {
		fp_dscv_devs_free(ddevs);
		return PAM_AUTHINFO_UNAVAIL;
	}

	r = find_dev_and_print(ddevs, prints, &ddev, &print);
	if (r) {
		fp_dscv_prints_free(prints);
		fp_dscv_devs_free(ddevs);
		send_info_msg(pamh, "Could not locate any suitable fingerprints "
			"matched with available hardware.");
		return PAM_AUTHINFO_UNAVAIL;
	}

	dev = fp_dev_open(ddev);
	fp_dscv_devs_free(ddevs);
	if (!dev) {
		fp_dscv_prints_free(prints);
		return PAM_AUTHINFO_UNAVAIL;
	}

	finger = fp_dscv_print_get_finger(print);

	r = fp_print_data_from_dscv_print(print, &data);
	fp_dscv_prints_free(prints);
	if (r) {
		fp_dev_close(dev);
		return PAM_AUTHINFO_UNAVAIL;
	}

	r = do_verify(pamh, dev, data, finger);
	fp_print_data_free(data);
	fp_dev_close(dev);
	return r;
}

PAM_EXTERN int pam_sm_authenticate(pam_handle_t *pamh, int flags, int argc,
	const char **argv)
{
	const char *rhost = NULL;
	FILE *fd;
	char buf[5];
	const char *username;
	char *homedir;
	struct passwd *passwd;
	int r;

	pam_get_item(pamh, PAM_RHOST, (const void **)(const void*) &rhost);
	if (rhost != NULL && strlen(rhost) > 0) {
		/* remote login (e.g. over SSH) */
		return PAM_AUTHINFO_UNAVAIL;
	}

	r = pam_get_user(pamh, &username, NULL);
	if (r != PAM_SUCCESS)
		return PAM_AUTHINFO_UNAVAIL;

	passwd = getpwnam(username);
	if (!passwd)
		return PAM_AUTHINFO_UNAVAIL;

	homedir = strdup(passwd->pw_dir);

	/* a bit of a hack to make libfprint use the right home dir */
	r = setenv("HOME", homedir, 1);
	if (r < 0) {
		free(homedir);
		return PAM_AUTHINFO_UNAVAIL;
	}

	r = do_auth(pamh);
	free(homedir);
	return r;
}

PAM_EXTERN int pam_sm_setcred(pam_handle_t *pamh, int flags, int argc,
	const char **argv)
{
	return PAM_SUCCESS;
}

PAM_EXTERN int pam_sm_chauthtok(pam_handle_t *pamh, int flags, int argc,
	const char **argv)
{
	return PAM_SUCCESS;
}



src/pam_fprint_enroll.c
/*
 * pam_fprint_enroll: Enrolls finger and saves the result to disk
 * Copyright (C) 2007 Vasily Khoruzhick <anarsoul@gmail.com>
 * Copyright (C) 2007 Daniel Drake <dsd@gentoo.org>
 *
 * Based on code from libfprint example 
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>

#include <libfprint/fprint.h>

static const char *finger_names[] = {
	[LEFT_THUMB] = "Left Thumb",
	[LEFT_INDEX] = "Left Index Finger",
	[LEFT_MIDDLE] = "Left Middle Finger",
	[LEFT_RING] = "Left Ring Finger",
	[LEFT_LITTLE] = "Left Little Finger",
	[RIGHT_THUMB] = "Right Thumb",
	[RIGHT_INDEX] = "Right Index Finger",
	[RIGHT_MIDDLE] = "Right Middle Finger",
	[RIGHT_RING] = "Right Ring Finger",
	[RIGHT_LITTLE] = "Right Little Finger",
};


static struct fp_dscv_dev *discover_device(struct fp_dscv_dev **discovered_devs)
{
	struct fp_dscv_dev *ddev = NULL;
	int i;

	for (i = 0; ddev = discovered_devs[i]; i++) {
		struct fp_driver *drv = fp_dscv_dev_get_driver(ddev);
		printf("Found device claimed by %s driver\n",
			fp_driver_get_full_name(drv));
		return ddev;
	}

	return ddev;
}

static struct fp_print_data *enroll(struct fp_dev *dev, enum fp_finger finger)
{
	struct fp_print_data *enrolled_print = NULL;
	int r;	

	printf("You will need to successfully scan your %s %d times to "
		"complete the process.\n", finger_names[finger], fp_dev_get_nr_enroll_stages(dev));

	do {
		sleep(1);
		printf("\nScan your finger now.\n");

		r = fp_enroll_finger_img(dev, &enrolled_print, NULL);
		
		if (r < 0) {
			printf("Enroll failed with error %d\n", r);
			return NULL;
		}

		switch (r) {
		case FP_ENROLL_COMPLETE:
			printf("Enroll complete!\n");
			break;
		case FP_ENROLL_FAIL:
			printf("Enroll failed, something wen't wrong :(\n");
			return NULL;
		case FP_ENROLL_PASS:
			printf("Enroll stage passed. Yay!\n");
			break;
		case FP_ENROLL_RETRY:
			printf("Didn't quite catch that. Please try again.\n");
			break;
		case FP_ENROLL_RETRY_TOO_SHORT:
			printf("Your swipe was too short, please try again.\n");
			break;
		case FP_ENROLL_RETRY_CENTER_FINGER:
			printf("Didn't catch that, please center your finger on the "
				"sensor and try again.\n");
			break;
		case FP_ENROLL_RETRY_REMOVE_FINGER:
			printf("Scan failed, please remove your finger and then try "
				"again.\n");
			break;
		}
	} while (r != FP_ENROLL_COMPLETE);

	if (!enrolled_print) {
		fprintf(stderr, "Enroll complete but no print?\n");
		return NULL;
	}

	printf("Enrollment completed!\n\n");
	return enrolled_print;
}

int main(int argc, char *argv[])
{
	int r = 1, i;
	int next_option;
	enum fp_finger finger = RIGHT_INDEX;

	struct fp_dscv_dev *ddev;
	struct fp_dscv_dev **discovered_devs;
	struct fp_dev *dev;
	struct fp_print_data *data;
	
	const char * const short_options = "hf:";
	const struct option long_options[] = {
		{ "help", 0, NULL, 'h'},
		{ "enroll-finger", 1, NULL, 'f'},
		{ NULL, 0, NULL, 0}
	};
	
	do {
		next_option = getopt_long(argc, argv, short_options, long_options,
			NULL);
		switch (next_option) {
		case 'h':
			/* Printing usage */
			printf("Usage: %s options\n", argv[0]);
			printf("	-h	--help			Display this usage information.\n"
			   "	-f	--enroll-finger index	Enroll finger with index.\n\n");
			printf("	Valid indexes are:\n");
			for (i = LEFT_THUMB; i <= RIGHT_LITTLE; i++) {
				printf("	%d - %s\n", i, finger_names[i]);
			}
			exit(1);
			   
			break;
		case 'f':
			sscanf(optarg, "%d", &finger);
			if (finger < LEFT_THUMB || finger > RIGHT_LITTLE) {
				printf("%s: Invalid finger index.\n", argv[0]);
				printf("%s: Valid indexes are:\n", argv[0]);
				for (i = LEFT_THUMB; i <= RIGHT_LITTLE; i++) {
					printf("%s: %d - %s\n", argv[0], i, finger_names[i]);
				}
				exit(1);
			}
			break;
		case -1:	/* Done with options. */
			break;
		default:	/* Unexpected option */
			exit(1);
		}
	} while (next_option != -1);

	printf("This program will enroll your finger, "
		"unconditionally overwriting any selected print that was enrolled "
		"previously. If you want to continue, press enter, otherwise hit "
		"Ctrl+C\n");
	getchar();

	r = fp_init();
	if (r < 0) {
		fprintf(stderr, "Failed to initialize libfprint\n");
		exit(1);
	}

	discovered_devs = fp_discover_devs();
	if (!discovered_devs) {
		fprintf(stderr, "Could not discover devices\n");
		goto out;
	}

	ddev = discover_device(discovered_devs);
	if (!ddev) {
		fprintf(stderr, "No devices detected.\n");
		goto out;
	}

	dev = fp_dev_open(ddev);
	fp_dscv_devs_free(discovered_devs);
	if (!dev) {
		fprintf(stderr, "Could not open device.\n");
		goto out;
	}

	printf("Opened device. It's now time to enroll your finger.\n\n");
	data = enroll(dev, finger);
	if (!data)
		goto out_close;

	r = fp_print_data_save(data, finger);
	if (r < 0)
		fprintf(stderr, "Data save failed, code %d\n", r);

	fp_print_data_free(data);
out_close:
	fp_dev_close(dev);
out:
	fp_exit();
	return r;
}
